### PR TITLE
MODINVSTOR-827 Clean kafka topics on tenant delete

### PR DESCRIPTION
**Pull request by OleksiiKuzminov**
### Purpose/Overview:
There are a lot of unused topics that can be on envs with several tenants. To decrease pressure on consumers modules it’s nice to clean unused topics in Kafka.

### Approach:

Delete topics for the tenant on the tenant delete operation

### CODE REVIEW

**Comment by marcjohnson-kint** [comment]
My primary concern with this change is the impact it may have on consumers.

As I understand it, consuming modules are configured to automatically create a topic if it isn't present at the time the consumer is defined.

That leads me to think that consumers are expecting a topic to be there.

Would this change mean that modules consuming these topics have the topic removed from underneath them? What impact will that have on those modules?

**Comment by OleksiiKuzminov** [comment]
@marcjohnson-kint All consumers use subscription by pattern. In this case, there are no problems from mo-search side.

**Review by pfilippov-epam — APPROVED** [review]
Looks good.

File: src/main/java/org/folio/rest/impl/TenantRefAPI.java [hunk 1]
```diff
@@ -154,8 +159,19 @@ static boolean isNew(TenantAttributes attributes, String featureVersion) {
     return future.compose(result -> runJavaMigrations(attributes, vertxContext, headers).map(result));
   }
 
+  @Validate
+  public void postTenant(TenantAttributes tenantAttributes, Map<String, String> headers,
+                         Handler<AsyncResult<Response>> handler, Context context) {
+    // delete Kafka topics if tenant purged
+    Future.succeededFuture()
+      .compose(x -> tenantAttributes.getPurge()
```
Comment (julianladisch) [inline src/main/java/org/folio/rest/impl/TenantRefAPI.java:167]: `Future.succeededFuture().compose(x -> foo)` is a noop and should be replaced by `foo`.

Comment (julianladisch) [inline src/main/java/org/folio/rest/impl/TenantRefAPI.java:167]: A tenant can be disabled with purge=true or without purge=true: https://github.com/folio-org/okapi/blob/master/doc/guide.md#disabling

For me the intention of https://issues.folio.org/browse/MODINVSTOR-827 is to delete the kafka topics in both cases. Disabling without purge=true allows to upgrade to a new version without the need for keeping the kafka topics in the old version.

File: src/main/java/org/folio/services/kafka/topic/KafkaAdminClientService.java [hunk 2]
```diff
@@ -53,11 +53,34 @@ public KafkaAdminClientService(Supplier<KafkaAdminClient> clientFactory) {
...
+      } else {
+        log.error("Unable to delete topics", res.cause());
+      }
+      kafkaAdminClient.close().onComplete(closeResult -> {
+        if (closeResult.failed()) {
+          log.error("Failed to close kafka admin client", closeResult.cause());
+        }
+      });
     });
+    return result.future();
```
Comment (julianladisch) [inline src/main/java/org/folio/services/kafka/topic/KafkaAdminClientService.java:79]: Avoiding the Promise and using dedicated Future methods:
```
    return kafkaAdminClient.deleteTopics(topicsToDelete)
        .onSuccess(x -> log.info("Topics deleted successfully"))
        .onFailure(e -> log.error("Unable to delete topics", e))
        .eventually(x -> 
            kafkaAdminClient.close()
            .onFailure(e -> log.error("Failed to close kafka admin client", e)));
```

Creating and closing the AdminKafkaClient is duplicated code (createKafkaTopics) and can be extracted into a `withAdminKafkaClient` method:

```
  public Future<Void> deleteKafkaTopics(String tenantId, String environmentName) {
    List<String> topicsToDelete = readTopics()
      .map(topic -> qualifyName(topic, environmentName, tenantId))
      .map(NewTopic::getName)
      .collect(Collectors.toList());
    return withKafkaAdminClient(kafkaAdminClient -> kafkaAdminClient.deleteTopics(topicsToDelete))
        .onSuccess(x -> log.info("Topics deleted successfully"))
        .onFailure(e -> log.error("Unable to delete topics", e));
  }

  private <T> Future<T> withKafkaAdminClient(Function<KafkaAdminClient, Future<T>> function) {
    final KafkaAdminClient kafkaAdminClient = clientFactory.get();
    return function.apply(kafkaAdminClient)
        .eventually(x ->
            kafkaAdminClient.close()
            .onFailure(e -> log.error("Failed to close kafka admin client", e)));
  }
```

File: src/main/java/org/folio/rest/impl/TenantRefAPI.java (hunk 1 above)
Comment (OleksiiKuzminov) [reply src/main/java/org/folio/rest/impl/TenantRefAPI.java]: @julianladisch  Can you please share an example of such requests for disabling without prge=true

**Review by MikhailFokanov — APPROVED** [review]
Looks good.

File: src/main/java/org/folio/rest/impl/TenantRefAPI.java (hunk 1 above)
Comment (OleksiiKuzminov) [reply src/main/java/org/folio/rest/impl/TenantRefAPI.java]: @julianladisch we decided to delete topics only on purge

**Comment by marcjohnson-kint** [comment]
@OleksiiKuzminov 

> All consumers use subscription by pattern. In this case, there are no problems from mo-search side.

What about other consumers other than mod-search?

Does that mean that the consumers don't automatically create the topics in the pattern themselves (and expect them to be there)?

Has mod-search been tested for tolerating the removal of a topic after consumption has started?

**Pull request merged**